### PR TITLE
Implement more performant swerve kinematics and odometry

### PR DESCRIPTION
**Pull request by MathNerd28**

### CODE REVIEW

**Comment by mmarchetti** [comment]
Is this based on WPILib swerve kinematics, or is it entirely new?

**Comment by MathNerd28** [comment]
It's essentially a clone (see [theirs](https://github.com/wpilibsuite/allwpilib/blob/main/wpimath/src/main/java/edu/wpi/first/math/kinematics/SwerveDriveKinematics.java)), with some performance optimizations and better method names. Additionally, I combined the Odometry and PoseEstimator into one class, since they're typically used together anyway.

File: src/main/java/org/victorrobotics/dtlib/subsystem/swerve/SwerveDrive.java [hunk 1]
```diff
@@ -114,29 +112,17 @@ public final void setStates(SwerveModuleState... states) {
...
-      double cosine = modules[i].getSteerAngle()
-                                .minus(states[i].angle)
-                                .getCos();
-      if (cosine < minCosine) {
-        minCosine = cosine;
-      }
-    }
-    minCosine *= minCosine * minCosine; // minCosine to the 3rd power
-    for (int i = 0; i < modules.length; i++) {
-      states[i].speedMetersPerSecond *= minCosine;
```
Comment (mmarchetti) [inline src/main/java/org/victorrobotics/dtlib/subsystem/swerve/SwerveDrive.java:129]: IIRC we originally added a correction like this to limit the speed of wheels that are facing the wrong way, until the motors align them a bit more with the direction of travel (though I don't recall the cubing). We'll need to do some testing as soon as we have a swerve chassis and shake all the issues out of this implementation before the build season.

**Review by mmarchetti — APPROVED** [review]

File: src/main/java/org/victorrobotics/dtlib/subsystem/swerve/SwerveDrive.java (hunk 1 above)
Comment (MathNerd28) [reply src/main/java/org/victorrobotics/dtlib/subsystem/swerve/SwerveDrive.java:129]: That's my plan for the new swerve chassis we just got

**Pull request merged**